File: Java/Sorting/QuickSort.java
public class QuickSort {

    private static void printArray(int[] array) {
        System.out.print("[");
        for (int i = 0; i < array.length - 1; i++) {
            System.out.print(array[i] + ", ");
        }
        System.out.println(array[array.length - 1] + "]");
    }

    private static int partition(int[] array, int start, int end) {
        int pivot = array[end];
        int index = start - 1;

        for (int j = start; j < end; j++) {
            if (array[j] <= pivot) {
                index++;

                int temp = array[index];
                array[index] = array[j];
                array[j] = temp;
            }
        }

        int temp = array[index + 1];
        array[index + 1] = array[end];
        array[end] = temp;

        printArray(array);

        return index + 1;
    }

    private static void quickSort(int[] array, int start, int end) {
        if (start < end) {
            int partitionIndex = partition(array, start, end);

            quickSort(array, start, partitionIndex - 1);
            quickSort(array, partitionIndex + 1, end);
        }
    }

    public static void main(String[] args) {
        int[] unsortedArray  = {9, 3, 2, 7, 1, 4, 5, 8, 6};
        System.out.println("Unsorted Array: ");
        printArray(unsortedArray);

        System.out.println("\nQuick Sort:");
        quickSort(unsortedArray, 0, unsortedArray.length - 1);
    }

}

File: C++/sorting/quick_sort.cpp
#include <iostream>
#include <vector>

int partition(std::vector<int> &array, int start, int end);
void quick_sort(std::vector<int> &array, int stat, int end);
void print_array(std::vector<int> array);

int partition(std::vector<int> &array, int start, int end)
{
    int pivot = array[end];
    int i = start - 1, j;

    for (j = start; j < end; j++)
    {
        if (array[j] <= pivot)
        {
            i++;

            std::swap(array[i], array[j]);
        }
    }

    std::swap(array[i + 1], array[end]);

    print_array(array);

    return i + 1;
}

void quick_sort(std::vector<int> &array, int start, int end)
{
   if (start < end)
   {
        int partition_index = partition(array, start, end);

        quick_sort(array, start, partition_index - 1);
        quick_sort(array, partition_index + 1, end);
    }
}

void print_array(std::vector<int>array)
{
    std::cout << "[";
    for(size_t i = 0; i < array.size() - 1; i++)
    {
        std::cout << array[i] << ", ";
    }
    std::cout <<  array[array.size() - 1] << "]" << std::endl;
}

int main(int argc, const char *argv[])
{
    std::vector<int> unsorted_array = {9, 3, 2, 7, 1, 4, 5, 8, 6};
    printf("Unsorted Array:\n");
    print_array(unsorted_array);

    printf("\nQuick Sort:\n");
    quick_sort(unsorted_array, 0, unsorted_array.size() - 1);

    return 0;
}
